File: src/main/java/com/server/ordering/FirebaseCloudMessageService.java
package com.server.ordering;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.auth.oauth2.GoogleCredentials;
import com.google.common.net.HttpHeaders;
import com.server.ordering.domain.OrderType;
import lombok.RequiredArgsConstructor;
import okhttp3.*;
import org.springframework.core.io.ClassPathResource;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.util.List;
import java.util.Objects;

@Component
@RequiredArgsConstructor
public class FirebaseCloudMessageService {

    private final ObjectMapper objectMapper;

    public void sendMessageTo(String targetToken, String title, String body, OrderType channel) throws IOException {

        String message = makeMessage(targetToken, title, body, channel);

        OkHttpClient client = new OkHttpClient();
        RequestBody requestBody = RequestBody.create(message, MediaType.get("application/json; charset=utf-8"));
        String API_URL = "https://fcm.googleapis.com/v1/projects/ordering-f880f/messages:send";
        Request request = new Request.Builder()
                .url(API_URL)
                .post(requestBody)
                .addHeader(HttpHeaders.AUTHORIZATION, "Bearer " + getAccessToken())
                .addHeader(HttpHeaders.CONTENT_TYPE, "application/json; UTF-8")
                .build();

        Response response = client.newCall(request)
                .execute();

        System.out.println(Objects.requireNonNull(response.body()).string());
    }

    private String makeMessage(String targetToken, String title, String body, OrderType channel) throws JsonProcessingException {

        FcmMessage fcmMessage = FcmMessage.builder()
                .message(FcmMessage.Message.builder()
                        .token(targetToken)
                        .data(FcmMessage.Data.builder()
                                .title(title)
                                .body(body)
                                .channel_id(channel.name())
                                .build())
                        .build())
                .validate_only(false)
                .build();
        return objectMapper.writeValueAsString(fcmMessage);
    }

    private String getAccessToken() throws IOException {

        String firebaseConfigPath = "firebase/firebase_service_key.json";

        GoogleCredentials googleCredentials = GoogleCredentials.fromStream(new ClassPathResource(firebaseConfigPath).getInputStream())
                .createScoped(List.of("https://www.googleapis.com/auth/cloud-platform"));

        googleCredentials.refreshIfExpired();

        return googleCredentials.getAccessToken().getTokenValue();
    }
}

File: src/main/kotlin/com/byeoru/ordering_server/FirebaseCloudMessageService.kt
package com.byeoru.ordering_server

import com.byeoru.ordering_server.domain.OrderType
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.google.auth.oauth2.GoogleCredentials
import com.google.common.net.HttpHeaders
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.springframework.core.io.ClassPathResource
import org.springframework.stereotype.Component
import java.io.IOException

@Component
class FirebaseCloudMessageService(private val objectMapper: ObjectMapper) {

    @Throws(IOException::class)
    fun sendMessageTo(targetToken: String, title: String, body: String, channel: OrderType) {
        val message = makeMessage(targetToken, title, body, channel)
        val client = OkHttpClient()
        val requestBody: RequestBody = message.toRequestBody("application/json; charset=utf-8".toMediaType())
        val API_URL = "https://fcm.googleapis.com/v1/projects/ordering-f880f/messages:send"
        val request: Request = Request.Builder()
            .url(API_URL)
            .post(requestBody)
            .addHeader(HttpHeaders.AUTHORIZATION, "Bearer $accessToken")
            .addHeader(HttpHeaders.CONTENT_TYPE, "application/json; UTF-8")
            .build()

        client.newCall(request).execute()
    }

    @Throws(JsonProcessingException::class)
    private fun makeMessage(targetToken: String, title: String, body: String, channel: OrderType): String {
        val fcmMessage = FcmMessage(
            message = FcmMessage.Message(
                data = FcmMessage.Data(
                    title = title,
                    body = body,
                    channel_id = channel.name
                ),
                token = targetToken
            ),
            validate_only = false
        )
        return objectMapper.writeValueAsString(fcmMessage)
    }

    @get:Throws(IOException::class)
    private val accessToken: String
        get() {
            val firebaseConfigPath = "firebase/firebase_service_key.json"
            val googleCredentials = GoogleCredentials.fromStream(ClassPathResource(firebaseConfigPath).inputStream)
                .createScoped(listOf("https://www.googleapis.com/auth/cloud-platform"))
            googleCredentials.refreshIfExpired()
            return googleCredentials.accessToken.tokenValue
        }
}